compiler/rustc_mir_transform/src/check_finalizers.rs

```diff
@@ -193,29 +200,64 @@ impl<'tcx> FinalizationCtxt<'tcx> {
                         tys.push(f)
                     }
                 }
-                ty::Adt(def, substs) if !self.is_copy(ty) => {
-                    for field in def.all_fields() {
-                        let field_ty = self.tcx.type_of(field.did).instantiate(self.tcx, substs);
-                        tys.push(field_ty);
-                    }
-
-                    if !def.has_dtor(self.tcx) {
-                        continue;
-                    }
-
+                ty::Adt(def, substs) => {
                     if def.is_box() {
                         // This is a special case because Box has an empty drop
                         // method which is filled in later by the compiler.
                         errors.push(FinalizerErrorKind::MissingFnDef);
                     }
-                    DropMethodChecker::new(ty, self, &mut errors).check();
+                    let mut res = if !def.has_dtor(self.tcx) || self.is_copy(ty) {
```

Comment:
This logic feels a bit odd to me: `res` isn't assigned to again AFAICS? This `if` and the `match` look like they could probably be condensed to something like:

```rust
if def.has_tor(...)  !self.is_copy(ty) {
  match DropMethodChecker::new(...) {
    // as before
  }
}
```

?

Reply:
That's not quite what's happening. We match on  `res` later on to determine whether or not to add the `Adt`'s fields to the stack. However, if `def` does not have a destructor, we still need to go through its fields. What about something like this (58b8194c7220864956cf949253e48d40655b2dc5)? There's a lot of nested conditionals happening now but it might be a bit clearer than before.

Reply:
Yes, that's clearer!